### PR TITLE
fix(NODE-3153): correctly deserialize `__proto__` properties

### PR DESCRIPTION
**Pull request by addaleax**
NODE-3153

## Description

**What changed?**

### CODE REVIEW

**Review by emadum — APPROVED** [review]

File: src/parser/deserializer.ts [hunk 1]
```diff
@@ -599,12 +599,22 @@ function deserializeObject(
...
       );
     }
+    if (name === '__proto__') {
+      Object.defineProperty(object, name, {
+        value,
+        writable: true,
+        enumerable: true,
+        configurable: true
+      });
+    } else {
```
Comment (dariakp) [inline src/parser/deserializer.ts:615]: What's the reason for having an else for the non-proto cases? Are there special properties that the above block would not handle in the way we want?

Comment (addaleax) [reply src/parser/deserializer.ts:615]: @dariakp This was a performance concern that @nbbeeken and I talked about in the ticket – I did benchmark this, `Object.defineProperty()` is (somewhat understandably) significantly slower than just setting the property directly.

Comment (dariakp) [reply src/parser/deserializer.ts:615]: Oh that's interesting, I saw the note on the ticket saying that we weren't sure whether it was optimized by V8, but not the actual outcome of any benchmarking. Just out of curiosity, what *was* the performance degradation?

Comment (addaleax) [reply src/parser/deserializer.ts:615]: It’s about 50 % slower: https://jsbench.me/w7ko1ntaf5/1

I guess it’s fair that V8 doesn’t optimize this as much, it’s not nearly as common for properties to be defined this way.

**Review by dariakp — APPROVED** [review]

File: src/parser/deserializer.ts (hunk 1 above)
Comment (dariakp) [reply src/parser/deserializer.ts:615]: Fair and yeah, that certainly qualifies as significant, thanks for providing the context!

**Pull request merged**